### ISSUE TITLE
Replay and condition same variable

### ISSUE THREAD

**Issue by sharrison5** [ratschlab/projects2019-boosting-bbvi-pyro#2]
https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/blob/aa485ae1db251478d5d49c864d7728b3eb6a37e0/prototype.py#L43
https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/blob/aa485ae1db251478d5d49c864d7728b3eb6a37e0/prototype.py#L55

Just been having a play with the code, and I think the two lines above might be an issue - I don't think that we can simultaneously condition via `obs=` and replay via `guide()` for the same variable. My mental model is that these are trying to do the same thing: replace whatever is sampled from the prior with either (1) the data or (2) a sample from the posterior. But I'm not sure if it's meaningful / possible for that to happen simultaneously: only one can actually be used when you call `trace.log_prob_sum()` (AFAIK - a quick skim of minipyro suggests one would overwrite the other at `msg["value"]`).

**Comment by lorenzkuhn** [comment]
Hmm, good point - I'll look into it!

**Comment by lorenzkuhn** [comment]
Ok, I had a look. I think you're right, only one of those two things can happen. Looking at the Pyro source code, it looks like a sample site's value will not be changed by `replay`, if we registered an observation for the sample:

https://github.com/pyro-ppl/pyro/blob/06f98782ead0262a5f8d83e439dd469b8238d6b6/pyro/poutine/replay_messenger.py#L36

I've also had another look at the [Pyro Gaussian Mixture Models tutorial](http://pyro.ai/examples/gmm.html) for density estimation. It seems like they condition the samples in the model on the observations just like we did. Let's discuss this tomorrow!
